### ISSUE TITLE
Memory in SpaceTimeStack tool seems wrong

### ISSUE THREAD

**Issue by stanmoore1** [kokkos/kokkos-tools#20]
The memory in the `SpaceTimeStack` tool seems wrong. The high water mark from this tool (MAX BYTES ALLOCATED) doesn't match the other Kokkos memory high water mark tools, and the percentages listed don't add up to 100%.

@ibaned

**Comment by stanmoore1** [comment]
Also, I think that there needs to be a separate high water mark for CUDA and Host memory spaces.

**Comment by ibaned** [comment]
The not adding up to 100% is probably due to a lower cutoff threshold that omits entries with less than a few percent contribution. I'll need a reproducer of some kind to debug the difference in totals between tools.

**Comment by ibaned** [comment]
also, the `hwm` tools use this call:
```
getrusage(RUSAGE_SELF, &sys_resources);
sys_resources.ru_maxrss
```
which is an estimate of max CPU memory used by any memory allocation calls.
`space-time-stack` only measures `Kokkos::View` allocations, nothing more.

**Comment by stanmoore1** [comment]
Here is the output from space-time-stack. The host memory only adds up to 19.40% and the CUDA memory adds up to 79.70%. Are those two supposed to make 100% together? I don't think that makes sense since host and CUDA use totally separate memory.

<pre>
MPI RANK WITH MAX MEMORY: 0
MAX BYTES ALLOCATED: 10696989988
HOST ALLOCATIONS:
================
  15.5% particle:particles_mirror
  1.8% grid:cells_mirror
  1.0% grid:cinfo_mirror
  0.6% particle:mlist_mirror
  0.1% emit/face:tasks_mirror
  0.1% collide:vremax_mirror
  0.1% collide:remain_mirror
  0.1% thermal/grid:vector_grid_mirror
  0.1% thermal/grid:tally_mirror
CUDA ALLOCATIONS:
================
  15.5% particle:particles
  15.5% particle:sorted
  15.4% particle:particles
  10.4% particle:plist
  9.8% particle:plist
  9.2% particle:plist
  1.8% grid:cells
  1.0% grid:cinfo
  0.6% particle:mlist
  0.1% emit/face:tasks
  0.1% collide:vremax
  0.1% collide:remain
  0.1% thermal/grid:vector_grid
  0.1% thermal/grid:tally

END KOKKOS PROFILING REPORT.
</pre>

**Comment by ibaned** [comment]
Yes, it does treat them both as one total. I agree it makes more sense to separate the totals. I'll treat this issue as a request to do that.

**Comment by ibaned** [comment]
I'll also bundle into this the separate calculation of high water marks for each space.

**Comment by stanmoore1** [comment]
@ibaned thanks. 

Using the `memory-usage` tool, I get a high water mark of 3595.9 MB for host and 8773.9 MB for CUDA, which adds up to 12369.8 MB total memory in Kokkos views, or 1.297e10 bytes. It doesn't quite match `MAX BYTES ALLOCATED: 10696989988` from `space-time-stack`, any idea why?

**Comment by ibaned** [comment]
The `HighWater(MB)` column is based on Kokkos allocation events, while the `HighWater-Process(MB)` is based on `getrusage`. The only one I would expect to match with `space-time-stack` is `HighWater(MB)`.  Also this seems badly named, it is really *current* memory usage by Kokkos Views, while `HighWater-Process(MB)` is actually the high water value from `getrusage`.

**Comment by stanmoore1** [comment]
Actually no, `Size(MB)` is the current memory usage from Kokkos views, `HighWater(MB)` is the high water mark from Kokkos views, and `HighWater-Process(MB)` is the actual high water mark *on the host*. The last value shouldn't be included in the CUDA output since it is for the host, see #21.

The numbers I posted are from `HighWater(MB)`  and they should match the `space-time-stack` high water mark as far as I can tell, but don't.

**Comment by ibaned** [comment]
oh, yep, I read the code wrong with regards to `Size(MB)` and `HighWater(MB)`.

I'll try it on some random Kokkos app when I get a chance....

**Comment by stanmoore1** [comment]
>Using the memory-usage tool, I get a high water mark of 3595.9 MB for host and 8773.9 MB for CUDA, which adds up to 12369.8 MB total memory in Kokkos views, or 1.297e10 bytes. It doesn't quite match MAX BYTES ALLOCATED: 10696989988 from space-time-stack, any idea why?

I think this is due to the non-determinism in my program.

**Comment by ibaned** [comment]
The change to separate Host and CUDA high water marks is merged into master via PR #23.
